cpp/src/hash/hashing.cu

```diff
@@ -77,84 +82,394 @@ class bitwise_partitioner
+    } else {
+      temp_histo[i] = 0;
+    }
+  }
+
+  __syncthreads();
+
+  BlockScan(temp_storage).ExclusiveSum(temp_histo, temp_histo);
+
+  __syncthreads();
+
+  if (threadIdx.x == 0) {
+    partition_offset_shared[0] = 0;
+  }
+
+  for (int i = 0; i < ELEMENT_PER_BLOCK; ++i) {
+    if (ELEMENT_PER_BLOCK * threadIdx.x + i < num_partitions) {
+      partition_offset_shared[ELEMENT_PER_BLOCK * threadIdx.x + i + 1] = temp_histo[i]; 
+    } 
+  }
+
+  __syncthreads();
+
+  // Fetch the offset in the output buffer of each partition in this thread block
+  ipartition = threadIdx.x;
+  while (ipartition < num_partitions) {
+    partition_offset_global[ipartition] = scanned_block_partition_sizes[ipartition * gridDim.x + blockIdx.x];
+    ipartition += blockDim.x;
+  }
+
+  __syncthreads();
+
+  size_type row_number = threadIdx.x + blockIdx.x * blockDim.x;
+
+  // Fetch the input data to shared memory
+  while ( row_number < num_rows ) {
+    ipartition = row_partition_numbers[row_number];
+
+    block_output[
+      partition_offset_shared[ipartition] + row_partition_offset[row_number]
+    ] = input_buf[row_number];
+
+    row_number += blockDim.x * gridDim.x;
+  }
+
+  __syncthreads();
+
+  // Copy data from shared memory to output buffer
+
+  constexpr int nthreads_partition = 16; // Use 16 threads to copy each partition, assume BLOCK_SIZE
+                                         // is divisible by 16
+  
+  for (ipartition = threadIdx.x / nthreads_partition;
+       ipartition < num_partitions;
+       ipartition += BLOCK_SIZE / nthreads_partition) {
+    
+    size_type row_offset = threadIdx.x % nthreads_partition;
+    size_type nelements_partition = partition_offset_shared[ipartition + 1] - partition_offset_shared[ipartition];
+
+    while (row_offset < nelements_partition) {
+      output_buf[partition_offset_global[ipartition] + row_offset]
+        = block_output[partition_offset_shared[ipartition] + row_offset];
+      
+      row_offset += nthreads_partition;
+    }
+  }
+}
+
+struct move_to_output_buffer_dispatcher{
+  template <typename DataType,
+      std::enable_if_t<is_fixed_width<DataType>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& input,
+                  const size_type num_partitions,
+                  size_type * row_partition_numbers,
+                  size_type * row_partition_offset,
+                  size_type * block_partition_sizes,
+                  size_type * scanned_block_partition_sizes,
+                  size_type grid_size,
+                  rmm::mr::device_memory_resource* mr,
+                  cudaStream_t stream)
+  {
+    CUDF_EXPECTS(input.null_mask() == nullptr, "null input column unsupported");
+
+    rmm::device_buffer output(input.size() * sizeof(DataType), stream, mr);
+
+    int const smem = BLOCK_SIZE * ROWS_PER_THREAD * sizeof(DataType)
+      + (num_partitions + 1) * sizeof(size_type) * 2;
```

Comment:
Should there be a check that num_partitions is less than some reasonable value like 2048, otherwise the launch will likely fail (32KB per threadblock) ~ not sure what's a typical value for num_partitions.

Reply:
Currently my optimization only supports up to 1024 partitions but I should be able to support 2048 partitions. Waiting for others to comment maximum number of partitions we need to support

Reply:
I also have the question on maximum number of partitions we need to support. Currently my code only supports up to 1024 number of partitions but I can easily change the code to support 2048 partition